### ISSUE TITLE
Prometheus: format_query support

### ISSUE THREAD

**Issue by gtk-grafana** [grafana/grafana#68732]
### Discussed in https://github.com/grafana/grafana/discussions/68579

<div type='discussions-op-text'>

<sup>Originally posted by **dimitarvdimitrov** May 16, 2023</sup>
Prometheus supports the [`/api/v1/format_query` endpoint](https://prometheus.io/docs/prometheus/latest/querying/api/#formatting-query-expressions) which takes a promql expression and prettifies it. 

Recently this was added to Grafana Mimir too. See https://github.com/grafana/mimir/pull/4373

Grafana can use this endpoint to format the query in the editor.</div>

@catherineymgui what do you think about a "prettify" component in the Prometheus code editor?

**Comment by catherineymgui** [comment]
What does prettifying an expression mean? And what is the use case / customer pain point?

**Comment by gtk-grafana** [comment]
It will apply consistent formatting rules to promQL, which greatly aids when writing complex queries in the code editor, or pasting in promQL. It's also called auto-formatting or "Beautify", common UI feature when dealing with syntax heavy strings. 

https://jsonformatter.org/ or anything dealing with JSON will have this functionality

**Comment by gtk-grafana** [comment]
Actually this might not be as easy as I'd hoped, since Grafana is interpolating some variables on the backend, i.e. the $__rate_interval, we'll need to interpolate those to proper promQL on the backend and then back to Grafana's superset of promql, except the indices of the replacements in the unformatted string will not associate with the locations in the formatted query.

TL;DR this should be easy except Grafana query expressions aren't technically valid promQL because of the custom variables.

**Comment by gtk-grafana** [comment]
Closing, we'll need to find another way to format promQL within Grafana

**Comment by taleena** [comment]
@gtk-grafana is there another issue that's capturing where we're looking into this?

**Comment by gtk-grafana** [comment]
I don't believe so. @grafana/observability-metrics do we have an open issue for adding promQL formatting?

**Comment by taleena** [comment]
Reopening so we can make sure to follow up and get another issue linked. :)

**Comment by itsmylife** [comment]
Another approach we can use 
Details: https://raintank-corp.slack.com/archives/C02UC0KCSF4/p1716297150875969
Code: https://github.com/o11ydev/oy-toolkit/blob/main/wasm/promqlparser/main.go

**Comment by itsmylife** [comment]
I think I can handle this with the new language_provider implementation.
